sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/OrderedEventProcessor.java

```diff
@@ -0,0 +1,657 @@
+        // Only when the current event caused a state mutation and the state is passed to this
+        // method should we attempt to process buffered events
+        return;
+      }
+
+      if (!processingState.readyToProcessBufferedEvents()) {
+        return;
+      }
+
+      if (exceededMaxResultCountForBundle(processingState, largeBatchEmissionTimer)) {
+        // No point in trying to process buffered events
+        return;
+      }
+
+      Instant startRange = Instant.ofEpochMilli(processingState.getEarliestBufferedSequence());
+      Instant endRange = Instant.ofEpochMilli(processingState.getLatestBufferedSequence() + 1);
+      Instant endClearRange = null;
+
+      // readRange is efficiently implemented and will bring records in batches
+      Iterable<TimestampedValue<EventTypeT>> events =
+          bufferedEventsState.readRange(startRange, endRange);
+
+      Iterator<TimestampedValue<EventTypeT>> bufferedEventsIterator = events.iterator();
+      while (bufferedEventsIterator.hasNext()) {
+        TimestampedValue<EventTypeT> timestampedEvent = bufferedEventsIterator.next();
+        Instant eventTimestamp = timestampedEvent.getTimestamp();
+        long eventSequence = eventTimestamp.getMillis();
+
+        EventTypeT bufferedEvent = timestampedEvent.getValue();
+        if (processingState.checkForDuplicateBatchedEvent(eventSequence)) {
+          outputReceiver
+              .get(unprocessedEventsTupleTag)
+              .output(
+                  KV.of(
+                      processingState.getKey(),
+                      KV.of(
+                          eventSequence,
+                          UnprocessedEvent.create(bufferedEvent, Reason.duplicate))));
+          continue;
+        }
+
+        if (eventSequence > processingState.getLastOutputSequence() + 1) {
+          processingState.foundSequenceGap(eventSequence);
+          // Records will be cleared up to this element
+          endClearRange = Instant.ofEpochMilli(eventSequence);
+          break;
+        }
+
+        // This check needs to be done after we checked for sequence gap and before we
+        // attempt to process the next element which can result in a new result.
+        if (exceededMaxResultCountForBundle(processingState, largeBatchEmissionTimer)) {
+          endClearRange = Instant.ofEpochMilli(eventSequence);
+          break;
+        }
+
+        state.mutate(bufferedEvent);
+        ResultTypeT result = state.produceResult();
+        if (result != null) {
+          outputReceiver.get(mainOutputTupleTag).output(KV.of(processingState.getKey(), result));
+          processingState.resultProduced();
+        }
+        processingState.processedBufferedEvent(eventSequence);
+        // Remove this record also
+        endClearRange = Instant.ofEpochMilli(eventSequence + 1);
+      }
+
+      bufferedEventsState.clearRange(startRange, endClearRange);
+    }
+
+    private boolean exceededMaxResultCountForBundle(
```

Comment:
nit, maybe rename to `reachedMaxResultCountForBundle` since the check is `>=`

Reply:
Fixed